### PR TITLE
Next

### PR DESCRIPTION
**Pull request by stefannegrea**

### CODE REVIEW

**Comment by stefannegrea** [comment]
I opened the PR to get early feedback.

**Comment by burmanm** [comment]
Confusing, why Guice? Google has killed the project basically. Dagger is what they're using these days.

**Comment by stefannegrea** [comment]
@burmanm, I was not aware of Dagger. I needed something compatible with javax.inject to bootstrap the legacy `MetricsServiceLifecycle` class (which itself needs refactoring). 

Once things get more stable I will try to port the Guice code to Dagger.

**Comment by jsanda** [comment]
Do we need to port MetricsServiceLifecycle? That class is a total mess. We obviously need to initialize the various components and services, but DI service should handle most of that. 

Dagger does support the javax.inject annotations. I am not sure it makes sense to invest in something, i.e., Guice, that we know we will move away from.

**Comment by stefannegrea** [comment]
@jsanda, the functionality from `MetricsServiceLifecycle` needs to be ported somewhere else because it does all the setup. 

My goal for the moment is to port an entire handler class to be able to run some performance testing. For now I kept `MetricsServiceLifecycle` mostly unchanged and use Guice to do the work WF was doing with javax.inject annotations. Since the port of the handler is more involved, I try to keep non-handler code (eg. `MetricsServiceLifecycle`) as-is until we run some performance testing.

**Comment by jsanda** [comment]
As per our discussion at the end of last week, my personal and strong preference in general is to keep PRs small. There are several benefits such as making it easier to review, making it easier for others to get involved as changes land in master soon, and reducing the chances of feature creep. Sometimes it is relatively easy to break work down into smaller PRs. Other times it is more difficult. 

I would like to see the following for an initial PR that starts the migration to Vert.x:

* An endpoint for ingesting metrics (for a single metric type)
* An endpoint for reading raw metrics (for a single metric type)
* Automated tests

That's it, nothing more. It makes it much easier to review, easier to focus on the changes in question since the scope is more narrow, and easier to iterate on those changes. The tenant filter, wiring up and initializing components, etc., can be done in subsequent tickets.

**Comment by jsanda** [comment]
I just looked at MetricHandler.java. None of the query endpoints stream the response. One of the big motivations for moving off of JAX-RS is so that we can do streaming more easily. Rather than trying to port a whole bunch of functionality, I would prefer to see one set of read/write endpoints that is done taking a more async+streaming approach.

File: api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/util/Wrappers.java [hunk 1]
```diff
@@ -75,4 +79,20 @@ public static void setupTenantRoute(Router router, HttpMethod method, String pat
                 },
                 ctx.response()::end);
     }
+
+    public static Transformer<String, String> toJsonArray(RoutingContext ctx) {
```
Comment (burmanm) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/util/Wrappers.java]: No no, nothing like this. Real JSON parser (like the example I gave you). This one tests nothing useful for us.

Comment (jsanda) [inline api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/util/Wrappers.java]: @burmanm Can you provide a link/reference to that example? I don't see it mentioned anywhere in the PR comments.

**Comment by jsanda** [comment]
I got back to reviewing this branch a little bit this evening, and I have a couple observations. First, the vert.x implementation appears to be packaged as a WAR. Does vert.x even support WAR files? I did not think so, but maybe it does. That got me curious about how things are set up to run with the REST API integration tests. I discovered that the integration-tests module is commented out. This brings me to my question.

@stefannegrea how are you running the server in this branch?

My second observation is about dependencies. I peeked in the WEB-INF/lib directory and found the following:

```
antlr4-runtime-4.6.jar
aopalliance-1.0.jar
backport-util-concurrent-3.1.jar
cassalog-0.4.2.jar
cassandra-driver-core-3.0.5.jar
classworlds-1.1.jar
commons-cli-1.2.jar
commons-codec-1.10.jar
commons-exec-1.3.jar
commons-httpclient-3.0.jar
commons-lang3-3.3.1.jar
commons-math3-3.4.1.jar
compression-gorilla-1.1.0.jar
doxia-logging-api-1.1.jar
doxia-sink-api-1.1.jar
exec-maven-plugin-1.6.0.jar
groovy-all-2.4.6.jar
guava-18.0.jar
guice-4.1.0.jar
hawkular-commons-utils-0.9.7.Final-SRC-revision-b76074c26ebad20064f2a0b8185c5f6b1a11b415.jar
hawkular-cors-jaxrs-filter-0.9.7.Final-SRC-revision-b76074c26ebad20064f2a0b8185c5f6b1a11b415.jar
hawkular-metrics-api-util-0.28.0-SNAPSHOT.jar
hawkular-metrics-configuration-service-0.28.0-SNAPSHOT.jar
hawkular-metrics-core-service-0.28.0-SNAPSHOT.jar
hawkular-metrics-datetime-service-0.28.0-SNAPSHOT.jar
hawkular-metrics-job-scheduler-0.28.0-SNAPSHOT.jar
hawkular-metrics-model-0.28.0-SNAPSHOT.jar
hawkular-metrics-rx-java-driver-0.28.0-SNAPSHOT.jar
hawkular-metrics-schema-0.28.0-SNAPSHOT.jar
hawkular-rest-api-0.9.7.Final-SRC-revision-b76074c26ebad20064f2a0b8185c5f6b1a11b415.jar
hawkular-server-0.9.7.Final-SRC-revision-b76074c26ebad20064f2a0b8185c5f6b1a11b415.jar
infinispan-commons-8.0.1.Final.jar
infinispan-core-8.0.1.Final.jar
jackrabbit-jcr-commons-1.5.0.jar
jackrabbit-webdav-1.5.0.jar
jackson-annotations-2.5.4.jar
jackson-core-2.5.4.jar
jackson-databind-2.5.4.jar
jackson-dataformat-yaml-2.4.5.jar
jackson-datatype-joda-2.4.5.jar
jackson-jaxrs-base-2.6.3.jar
jackson-jaxrs-json-provider-2.6.3.jar
jackson-module-jaxb-annotations-2.6.3.jar
javax.inject-1.jar
jaxrs-api-3.0.12.Final.jar
jboss-logging-3.3.0.Final.jar
jboss-marshalling-osgi-1.4.10.Final.jar
jboss-transaction-api_1.1_spec-1.0.1.Final.jar
jcl-over-slf4j-1.5.6.jar
jgroups-3.6.4.Final.jar
jmh-core-1.13.jar
joda-time-2.9.5.jar
jopt-simple-4.6.jar
jsch-0.1.38.jar
maven-artifact-2.2.1.jar
maven-artifact-manager-2.2.1.jar
maven-core-2.2.1.jar
maven-error-diagnostics-2.2.1.jar
maven-model-2.2.1.jar
maven-monitor-2.2.1.jar
maven-plugin-api-2.2.1.jar
maven-plugin-descriptor-2.2.1.jar
maven-plugin-parameter-documenter-2.2.1.jar
maven-plugin-registry-2.2.1.jar
maven-profile-2.2.1.jar
maven-project-2.2.1.jar
maven-reporting-api-2.2.1.jar
maven-repository-metadata-2.2.1.jar
maven-settings-2.2.1.jar
maven-toolchain-2.2.1.jar
metrics-core-3.1.2.jar
nekohtml-1.9.6.2.jar
netty-buffer-4.1.8.Final.jar
netty-codec-4.1.8.Final.jar
netty-codec-dns-4.1.8.Final.jar
netty-codec-http2-4.1.8.Final.jar
netty-codec-http-4.1.8.Final.jar
netty-codec-socks-4.1.8.Final.jar
netty-common-4.1.8.Final.jar
netty-handler-4.0.37.Final.jar
netty-handler-proxy-4.1.8.Final.jar
netty-resolver-4.1.8.Final.jar
netty-resolver-dns-4.1.8.Final.jar
netty-transport-4.1.8.Final.jar
netty-transport-native-epoll-4.1.8.Final-linux-x86_64.jar
plexus-cipher-1.4.jar
plexus-container-default-1.0-alpha-9-stable-1.jar
plexus-interactivity-api-1.0-alpha-4.jar
plexus-interpolation-1.11.jar
plexus-sec-dispatcher-1.3.jar
plexus-utils-3.0.20.jar
resteasy-jackson2-provider-3.0.14.Final.jar
rxjava-1.2.9.jar
rxjava-guava-1.0.3.jar
rxjava-math-1.0.0.jar
slf4j-api-1.7.7.jar
slf4j-jdk14-1.5.6.jar
slf4j-nop-1.5.3.jar
snakeyaml-1.12.jar
swagger-annotations-1.5.8.jar
swagger-core-1.5.8.jar
swagger-models-1.5.8.jar
validation-api-1.1.0.Final.jar
vertx-auth-common-3.4.1.jar
vertx-core-3.4.1.jar
vertx-rx-java-3.4.1.jar
vertx-web-3.4.1.jar
wagon-file-1.0-beta-6.jar
wagon-http-1.0-beta-6.jar
wagon-http-lightweight-1.0-beta-6.jar
wagon-http-shared-1.0-beta-6.jar
wagon-provider-api-1.0-beta-6.jar
wagon-ssh-1.0-beta-6.jar
wagon-ssh-common-1.0-beta-6.jar
wagon-ssh-external-1.0-beta-6.jar
wagon-webdav-jackrabbit-1.0-beta-6.jar
xercesMinimal-1.9.6.2.jar
```
It looks like a lot of those are in support of vert.x code generation; however, there are still runtime dependencies, like JAX-RS that are in there. Both the Cassandra driver and Vert.x and built on netty, but on different versions. It looks like most of the netty dependencies are getting pulled from vert.x, but I see at least one getting pulled from the driver,  netty-handler-4.0.37.Final.jar. We cannot assume that the driver will work the version of netty used by vert.x and vice versa. Vert.x does have some class loader isolation so we should be able to pull in the version of netty required by the C* driver.

**Comment by stefannegrea** [comment]
@jsanda, really good question/observation about the C* and Vert.x requiring different netty versions, I did not consider it. If it was not netty, I think the answer would be simpler. I am going to do the research today. 

For delivery, today everything is packaged in a zip or tar.gz file. This api project will produce the core jar for the vertx implementation and then dependencies are in a subfolder. Unzip and run the `java -jar` command.

**Pull request closed without merging**